mbed_build/_internal/config/config_layer.py

```diff
@@ -43,16 +43,36 @@ def from_config_source(cls, config_source: ConfigSource, target_labels: List[str
 
         This method translates data found in ConfigSource into modifiers specific to the given target.
         """
-        modifiers_from_config = [
-            build_modifier_from_config_entry(key=key, data=data) for key, data in config_source.config.items()
-        ]
+        namespace = config_source.namespace
+        modifiers = []
+        for key, data in config_source.config.items():
+            modifiers.append(build_modifier_from_config_entry(key=_namespace(key, namespace), data=data))
 
-        modifiers_from_target_overrides: List[Callable] = []
         allowed_target_labels = ["*"] + target_labels
         for target_label, overrides in config_source.target_overrides.items():
             if target_label in allowed_target_labels:
-                modifiers_from_target_overrides.extend(
-                    build_modifier_from_target_override_entry(key=key, data=data) for key, data in overrides.items()
-                )
+                for key, data in overrides.items():
+                    modifiers.append(
+                        build_modifier_from_target_override_entry(key=_namespace(key, namespace), data=data)
+                    )
 
-        return cls(config_source=config_source, modifiers=(modifiers_from_config + modifiers_from_target_overrides))
+        return cls(config_source=config_source, modifiers=modifiers)
+
+
+def _namespace(key: str, namespace: str) -> str:
+    """Prefix configuration key with a namespace.
+
+    Namespace is ConfigSource wide, and is resolved at source build time.
+
+    It should be one of:
+    - "target"
+    - "application"
+    - library name (where "mbed_lib.json" comes from)
+
+    If given key is already namespaced, return it as is - this is going to be the case for
+    keys from "target_overrides" entries. Keys from "config" usually need namespacing.
+    """
+    if "." in key:
```

Comment:
Are we sure that there aren't any non-namespaced keys with '.' in?

Reply:
To be honest I never looked for those. https://github.com/ARMmbed/mbed-os/blob/64853b354fa188bfe8dbd51e78771213c7ed37f7/tools/config/__init__.py#L169 seems to indicate it'd be an invalid key. Again, validation vs parsing.

Reply:
It seems like an optimistic check to me. Would it be possible to get a list of available namespaces from the ConfigSources and see if the first part of a key split on '.' matches one of them and if not throw and error?  If for some reason the validation against including periods in keys names didn't work and it reached this point we could end up with some tricky bugs.

Reply:
I think this should be a separate step: config validation.

While it might be possible to _easily_ validate those names for `target.json` and `mbed_lib.json` sources, `mbed_app.json` is a bit odd.
1. `targets.json`:
- If key is not prefixed, prefix with `target`
- If key is prefixed, validate the prefix is `target`
2. `mbed_lib.json`:
- If key is not prefixed, prefix with `<name>` value from json
- If key is prefixed, validate the prefix is either `<name>` value or `target`
3. `mbed_app.json`:
- If key is not prefixed, prefix with `app`
- If key is prefixed, it's all good

I was trying to avoid adding validation logic and go with an assumption that given configuration is valid. Your comments make me think that maybe prefixing should happen in the `ConfigSource` and include some validation logic. Thoughts?

Reply:
To me it feels wrong to have a check for 'does it already have a key? If not, add one' but actually that check is just 'does it have a full stop in it?' which may or may not indicate it has a key. Whether or not there is validation elsewhere surely this isn't a great check?

Maybe you are right and the issue is that this is in slightly the wrong place and it should be part of the ConfigSource. It would certainly be easier for each source to check for the presence of a single '<key_name>.' rather than layer to have to check through all of them.

Reply:
What would make this check _great_? The prefix can be any string. Are you suggesting doing regex match on the key - something along the lines of `^.*\..*`?

Can you explain how it is easier for each source to check the presence of a single namespace vs layer checking all of them? I'm not understanding this bit.

Reply:
Oh I see, because the libs all have different names. Crikey there is no way to make config great again. 😭

Reply:
Ok, I am mollified. I hadn't quite groked that the namespace for each lib is just the name of the lib, I had simplified it in my head to one namespace per layer. Now I can see why there is no better way. I re-approve 👍